pkg/controller/ttlafterfinished/ttlafterfinished_controller.go

```diff
@@ -0,0 +1,446 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ttlafterfinished
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	batch "k8s.io/api/batch/v1"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	batchinformers "k8s.io/client-go/informers/batch/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	batchlisters "k8s.io/client-go/listers/batch/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/controller"
+	jobutil "k8s.io/kubernetes/pkg/controller/job"
+	"k8s.io/kubernetes/pkg/kubectl/scheme"
+	"k8s.io/kubernetes/pkg/util/metrics"
+)
+
+// Controller watches for changes of Jobs API objects. Triggered
+// by Job creation and updates, it enqueues Jobs that have finished and passed
+// the Jobs' `.spec.ttlSecondsAfterFinished` time period to the `jobsToCleanup`
+// queue. The Controller has workers who consume the
+// `jobsToCleanup` queue and send requests to the API server to delete the Jobs
+// accordingly.
+// This is implemented outside of Job controller for separation of concerns, and
+// because it will be extended to handle other finishable resource types.
+type Controller struct {
+	client   clientset.Interface
+	recorder record.EventRecorder
+
+	// jLister can list/get Jobs from the shared informer's store
+	jLister batchlisters.JobLister
+
+	// jStoreSynced returns true if the Job store has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	jListerSynced cache.InformerSynced
+
+	// Jobs that the controller will attempt to delete.
+	jobsToCleanup workqueue.RateLimitingInterface
+
+	// Jobs to be added to job cleanup workqueue later after it's ready for clean up.
+	// These jobs are stored in a thread-safe heap, and will be popped to job cleanup
+	// workqueue after their TTLs expire.
+	// Note: we cannot rely on workqueue.AddAfter, because it uses a buffered channel of
+	// limited size (1k), and the workqueue will be blocked after the channel is full.
+	jobsDelayCleanupQueueMutex sync.RWMutex
+	jobsDelayCleanupQueue      *cache.Heap
+
+	// The next time we should check the jobsDelayCleanupQueue.
+	nextCheckAtMutex sync.RWMutex
+	nextCheckAt      *time.Time
```

Comment:
Am I missing a place where we sleep for the diff between now and nextCheckAt, or a change in the delay queue?

Reply:
We cannot sleep because `nextCheckAt` might become earlier (e.g. when a Job that will expire earlier is added).

Reply:
you would get woken up by the new item being added to the queue, in golang terms:
```
select {
nextDelay : = delayqueuetop.addTime - now
case <- time.After(nextDelay):
case <- queueUpdatedCh:
}